File: BaseSDK/Windows/Samples/Base/DS86/SingleFrameDelayTest/SingleFrameDelayTest.cpp
#include <iostream>
#include <sys/timeb.h>
#include <thread>
#include <fstream>
#include <string>
#include "Scepter_api.h"
using namespace std;
int main()
{
	cout << "---SingleFrameDelayTest---" << endl;
	ScStatus status = SC_OTHERS;

	//SDK Initialize
	status = scInitialize();
	if (status != ScStatus::SC_OK)
	{
		cout << "scInitialize failed status:" << status << endl;
		system("pause");
		return -1;
	}

	uint32_t deviceCount = 0;
	ScDeviceInfo* pDeviceListInfo = NULL;
	ScDeviceHandle deviceHandle = 0;
	//1.Search and notice the count of devices.
	//2.get infomation of the devices. 
	//3.open devices accroding to the info.
	status = scGetDeviceCount(&deviceCount, 3000);
	if (status != ScStatus::SC_OK)
	{
		cout << "scGetDeviceCount failed! make sure pointer valid or called scInitialize()" << endl;
		system("pause");
		return -1;
	}
	cout << "Get device count: " << deviceCount << endl;
	if (0 == deviceCount)
	{
		cout << "scGetDeviceCount scans for 3000ms and then returns the device count is 0. Make sure the device is on the network before running the samples." << endl;
		system("pause");
		return -1;
	}

	pDeviceListInfo = new ScDeviceInfo[deviceCount];
	status = scGetDeviceInfoList(deviceCount, pDeviceListInfo);
	if (status == ScStatus::SC_OK)
	{
		if (SC_CONNECTABLE != pDeviceListInfo[0].status)
		{
			cout << "connect status: " << pDeviceListInfo[0].status << endl;
			cout << "The device state does not support connection." << endl;
			return -1;
		}
	}
	else
	{
		cout << "GetDeviceListInfo failed status:" << status << endl;
		return -1;
	}

	cout << "serialNumber:" << pDeviceListInfo[0].serialNumber << endl
		<< "ip:" << pDeviceListInfo[0].ip << endl
		<< "connectStatus:" << pDeviceListInfo[0].status << endl;

	status = scOpenDeviceBySN(pDeviceListInfo[0].serialNumber, &deviceHandle);
	if (status != ScStatus::SC_OK)
	{
		cout << "OpenDevice failed status:" << status << endl;
		return false;
	}
	// set soft work model
	status = scSetWorkMode(deviceHandle, SC_SOFTWARE_TRIGGER_MODE);
	if (status != ScStatus::SC_OK)
	{
		cout << "scSetWorkMode failed status:" << status << endl;
		return -1;
	}
	int frameRate = 0;
	status = scGetFrameRate(deviceHandle, &frameRate);
	if (status != ScStatus::SC_OK)
	{
		cout << "scGetFrameRate failed status:" << status << endl;
		return -1;
	}
	else
	{
		cout << "scGetFrameRate : " << frameRate << endl;
	}
	//Starts capturing the image stream
	status = scStartStream(deviceHandle);
	if (status != ScStatus::SC_OK)
	{
		cout << "scStartStream failed status:" << status << endl;
		return -1;
	}
	//Wait for the device to upload image data
	ScFrameReady FrameReady = { 0 };
	ScFrame depthFrame = { 0 };
	ofstream csvWriter;
	string fileName = "";
	csvWriter.open(fileName.append("SingleFrameDelayTest.csv"));
	if (!csvWriter.is_open())
	{
		cout << "csv file open failed"<< endl;
		return -1;
	}
	csvWriter << "frameIndex,TotalDelay,ExcludeDelayofExposure" << endl;
	//1.software trigger.
	//2.ReadNextFrame.
	//3.GetFrame acoording to Ready flag and Frametype.
	//4.sleep 1000/frameRate (ms)
	uint64_t endTimestamp = 0;
	uint64_t startTimestamp = 0;
	uint64_t deviceTimestamp = 0;
	uint64_t frameInterval = 0;
	uint64_t frameIntervalNTP = 0;

	cout << "Please input the number of tests:" << endl;
	uint32_t number = 0;
	cin >> number;
	for (int i = 0; i < number; i++)
	{
		timeb timeStart, timeEnd;
		//call the below api to trigger one frame, then the frame will be sent
		// if do not call this function, the frame will not be sent and the below call will return timeout fail
		status = scSoftwareTriggerOnce(deviceHandle);
		ftime(&timeStart); // record the start timestamp
		if (status != ScStatus::SC_OK)
		{
			cout << "scSoftwareTriggerOnce failed status:" << status << endl;
			continue;
		}
		//If no image is ready within 1200ms, the function will return ScRetGetFrameReadyTimeOut
		status = scGetFrameReady(deviceHandle, 1200, &FrameReady);
		if (status != ScStatus::SC_OK)
		{
			cout << "scGetFrameReady failed status:" << status << endl;
			continue;
		}
		if (1 == FrameReady.depth)
		{
			status = scGetFrame(deviceHandle, SC_DEPTH_FRAME, &depthFrame);
			if (depthFrame.pFrameData != NULL)
			{
				if (depthFrame.frameIndex % 10 == 0)
				{
					cout << "scGetFrame,status:" << status << "  "
						<< "frameType:" << depthFrame.frameType << "  "
						<< "frameIndex:" << depthFrame.frameIndex << endl;
				}
				ftime(&timeEnd);// record the end timestamp
				endTimestamp = timeEnd.time * 1000 + timeEnd.millitm;
				startTimestamp = timeStart.time * 1000 + timeStart.millitm;
				deviceTimestamp = depthFrame.deviceTimestamp;
				frameInterval = endTimestamp - startTimestamp;
				frameIntervalNTP = endTimestamp - deviceTimestamp;

				if (frameIntervalNTP > 2000)
				{
					csvWriter << depthFrame.frameIndex << "," << frameInterval << ",N/A" << endl;
				}
				else
				{
					csvWriter<< depthFrame.frameIndex  << "," << frameInterval << "," << frameIntervalNTP <<endl;
				}
			}
		}
		this_thread::sleep_for(chrono::milliseconds(1000 / frameRate));
	}
	csvWriter.close();

	//set slave false
	status = scSetWorkMode(deviceHandle, SC_ACTIVE_MODE);
	if (status != ScStatus::SC_OK)
	{
		cout << "scSetWorkMode failed status:" << status << endl;
		return -1;
	}

	//Stop capturing the image stream
	status = scStopStream(deviceHandle);
	if (status != ScStatus::SC_OK)
	{
		cout << "scStopStream failed status:" << status << endl;
		return -1;
	}

	//1.close device
	//2.SDK shutdown
	status = scCloseDevice(&deviceHandle);
	if (status != ScStatus::SC_OK)
	{
		cout << "scCloseDevice failed status:" << status << endl;
		return -1;
	}
	status = scShutdown();
	if (status != ScStatus::SC_OK)
	{
		cout << "scShutdown failed status:" << status << endl;
		return -1;
	}
	cout << "---end---" << endl;

	return 0;
}

File: MultilanguageSDK/Python/Samples/DS77/SingleFrameDelayTest/SingleFrameDelayTest.py
from pickle import FALSE, TRUE
import sys
sys.path.append('../../../')

from API.ScepterDS_api import *
import time

camera = ScepterTofCam()

print("---SingleFrameDelayTest---")
camera_count = camera.scGetDeviceCount(3000)
print("Get device count:", camera_count)
if camera_count <= 0:
    print("scGetDeviceCount scans for 3000ms and then returns the device count is 0. Make sure the device is on the network before running the samples.")
    exit()

device_info=ScDeviceInfo()

if camera_count > 0:
    ret,device_infolist=camera.scGetDeviceInfoList(camera_count)
    if ret==0:
        device_info = device_infolist[0]
    else:
        print('GetDeviceListInfo failed status:' + ret)
        exit()  
else: 
    print("there are no camera found")
    exit()

if  ScConnectStatus.SC_CONNECTABLE.value != device_info.status:
	print("connect statu:",device_info.status)  
	print("The device state does not support connection.")
	exit()
else:
    print("serialNumber: "+str(device_info.serialNumber))
    print("ip: "+str(device_info.ip))
    print("connectStatus: "+str(device_info.status))

ret = camera.scOpenDeviceBySN(device_info.serialNumber)
if  ret != 0:
    print('OpenDevice failed status:' + str(ret))
    exit()


ret = camera.scSetWorkMode(ScWorkMode.SC_SOFTWARE_TRIGGER_MODE)
if  ret != 0:  
    print("scSetWorkMode failed status:",ret)

ret, frameRate = camera.scGetFrameRate()
if ret == 0:
    print("frameRate :", frameRate)
else:
    print("scGetFrameRate :", ret)
    exit()

ret = camera.scStartStream()
if  ret != 0:
    print("scStartStream failed status:"+ str(ret))
    exit()

curPath = os.getcwd()
filename = curPath + "/SingleFrameDelayTest.csv"
try:
    file = open(filename, "w")
except IOError:
    print('csv file open failed')
    exit()
file.write("frameIndex,TotalDelay,ExcludeDelayofExposure\n")

'''
//1.software trigger.
//2.ReadNextFrame.
//3.GetFrame acoording to Ready flag and Frametype.
//4.sleep 1000/frameRate (ms)
'''
number = int(input('Please input the number of tests:'))
for i in range(number):
    ret = camera.scSoftwareTriggerOnce()
    startTime = int(time.time()*1000)
    if  ret != 0:  
        print("scSoftwareTriggerOnce failed:",ret)

    ret, frameready = camera.scGetFrameReady(c_uint16(1200))
    if  ret != 0:
        print("scGetFrameReady failed status:",ret)
        continue       
    if  frameready.depth:      
        ret,frame = camera.scGetFrame(ScFrameType.SC_DEPTH_FRAME)
        if  ret == 0:
            if 0 == frame.frameIndex % 10:
                print("scGetFrame,status:", ret, "  ", "frameType:", frame.frameType, "  ", "frameIndex:", frame.frameIndex)
            endTime = int(time.time()*1000)
            exposureEndTime = frame.hardwaretimestamp
            frameIntervalTime = endTime - startTime
            frameIntervalNtpTime = endTime - exposureEndTime
            if frameIntervalNtpTime > 2000:
                file.write("{0},{1},N/A\n".format(frame.frameIndex, frameIntervalTime))
            else:
                file.write("{0},{1},{2}\n".format(frame.frameIndex, frameIntervalTime, frameIntervalNtpTime))
    time.sleep(1/frameRate)
file.close()

ret = camera.scSetWorkMode(ScWorkMode.SC_ACTIVE_MODE)
if  ret != 0:  
    print("scSetWorkMode failed status:",ret)

ret = camera.scStopStream()
if  ret != 0:
    print("stopstream failed",ret)

ret = camera.scCloseDevice()     
if  ret != 0:
    print('scCloseDevice failed: ' + str(ret))
           